### DISCUSSION TITLE
[Proposal] String literals denoting matching substrings in list pattern

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#7576] [Language Ideas]
## Summary

An enhanced way to denote substrings when matching a string using list patterns.

## Motivation

Currently, using list patterns against strings for capturing a simple variable within a specific region is inconvenient, as you would have to specify each character individually. Take for example matching field names denoting backing fields for specific properties (they are in the form `<PropertyName>k__BackingField`). The only available syntax currently is this:
```csharp
name is ['<', .. string propertyName, '>', 'k', '_', '_', 'B', 'a', 'c', 'k', 'i', 'n', 'g', 'F', 'i', 'e', 'l', 'd']
```

Specifying each character individually defeats the purpose of using list patterns altogether, a better approach would be to use a combination of `StartsWith` and `EndsWith`, and defining a generic intermediate string capturing method [like so](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQ65IJwAoCiAbgIYA2ExwApgGICWVpAJgHLEC2VuARADzEAjAHwBrAPpiAQsQDGIulADm9Rk24BKdQG50mJAAYA/AAIiZCtRXM2nXFn3GoHKuvQBvdMa/GAwsQAOwBBgVADOAAqU1GBQuE6cADTGfNxJAPYQwMYkYMYKUFRgNlRJ3KIS0nIKygzMGjpo3sZwAOx5UAVFzg0AvrpoWABsxgJpaaS+AUEhEVGFsfbZ5iXNBsahwMRgwOmZq0bG7HRMTKQri1RQTK5oHo3eo+OHlDIAFmHGALxL5FQAdABlTbbUIAdTowFeuA2W2A6mMADIET8IP98FcwRCoZdrg0mkcTmcvo4IKRSHjvHQAGbGXDsF7vUI3Jp3JpNBRZNJUqmhKhZb4w7Z/AAyl0UkIpbPaWTOSkhxJIvxFYvlAFp1sDgMq5a9jOqcdrxa9JWyOcYcQBJK5UAAexK5PL5xgA1MZZUaTfjjqcqArlgBtB28rV/S3Wm0AXRNfXuUrj8c8Xlaz2AbzCvXQQA=).

## Design

Strings are very well-known types, and string literals have special casing in many a cases. In this one, a proposed way is to allow raw string literals inside a list pattern involving `IEnumerable<char>` types, including a `string`. With the above example in mind, the following should be legal:

```csharp
public static bool IsBackingFieldName(string name)
{
    return name is ['<', .. string propertyName, ">k__BackingField"];
}
```

This would mean that `name` starts with `'<'` and ends with `">k__BackingField"`, while also capturing the value in between the two affixes, into the variable `propertyName`. String literals inside a list pattern over `char` values would be handled with that in mind.

### Bonus

With the above use case in mind, and with the fact that most list patterns over char sequences are for this purpose, we could utilize the `StartsWith` and `EndsWith` methods that are present on the underlying type, if it's a `string` or a `[ReadOnly]Span<char>`. In other words, the above example would exactly be lowered into (code adjusted from the SharpLab example linked above):

```csharp
public static bool IsBackingFieldName(string name)
{
    bool matches = name.StartsWith('<') && name.EndsWith(">k__BackingField");
    string propertyName = null;
    if (matches)
    {
        int offset = 1;
        int length = name.Length - 1 - ">k__BackingField".Length;
        int endIndex = offset + length;
        propertyName = name[offset..endIndex];
    }
                       
   return matches;
}
```

## Alternatives

As demonstrated already:
- Usage of `StartsWith`, `EndsWith`, `Substring` by hand with some arithmetic operations on the indexes.
- Writing down each individual character in the list pattern as already available today.

## Credits
@just-ero for the example and the problem at hand

**Comment by jnm2** [comment]
Yes please, I've been running into this since near the beginning of using list patterns. We should consider `u8` strings too.

Syntax that came to mind first was additional constant _spreads_, rather than listing the string as an _element_ of the list pattern:
`['<', .. var propertyName, .. ">k__BackingField"]`

This would clear the way for constants to be used:
`[.. ConstPrefix, .. var contents, .. ConstSuffix]`

**Reply by Rekkonnect** [reply]
Nice observation about the usage of constants too. I wanted to avoid the usage of the spread operator in order to not confuse it with the actual spreads that are captured, however it seems that with constants we would run into conflicts as they could also represent types. I'd like to hear your thoughts on why you'd prefer the presence of the spread operator over its absence in this feature.

**Reply by jnm2** [reply]
Besides specifically wanting to use constants, it also feels more logical to me to bridge the two worlds using a spread. In `[.. const1, .. var middle, .. const2]`, you can see structurally that 'const1' and 'middle' are the same kinds of things: collections of elements, not directly elements themselves of the list being matched.

It also feels at first blush more safe for future generalizations beyond a special char/string or byte/u8 relationship.

On the other hand, it is interesting that `"a"` and `'a'` would mean the same thing, so the element of confusion would not be there.

**Reply by jnm2** [reply]
Imagine a future generalization that would enable the following progression.
Initially, you want to match any two characters:

```cs
[.. { Count: 2 }, .. var rest]
```

Now, you realize you want to match two specific characters:
```cs
[.. "xy", .. var rest]
```

Removing the `..` at that point seems like a strange thing to do.

(This generalization is exploring the logical consequences of choices made about whether `..` is used, but not assuming that we will or should actually implement this specific capability.)

**Reply by jnm2** [reply]
I also want to mention the equivalence I see between `['a', 'b', 'c']` and `"abc"`. As patterns on strings, these mean the same thing. There's even interest in letting collection expressions allow both forms to be the same string constant.

In collection expressions and also potentially (at least logically) in patterns, `[.. ['a, 'b', 'c'],` would indicate something different from `[['a', 'b', 'c'],`.
Therefore I'd expect `[.. "abc",` and `["abc"` to respectively correspond to each of those distinct meanings.

**Comment by Rekkonnect** [comment]
Posting as comment after revisiting and giving a fresh thought on this:

An alternative design can be the following:

```cs
name is $"<{string propertyName}>k__BackingField" 
```

This isn't quite inspired by VB's now-unsupported Like operator; this string de-interpolation pattern would be a more natural way to express and decompose string patterns. 

To formally describe the above syntax, we have a different kind of string interpolation, where the allowed interpolation parts can be either constant expressions or patterns.

The following limitations exist:
- The spread operator cannot be combined with a variable declaration
- A variable declaration **must** be any of the following types:
  - `char`, for a single character
  - `string`, for a substring (the spread operator isn't required)
  - Any other type that implements `IParsable<TSelf>`, with special support for `ISpanParsable<TSelf>` and `IUtf8SpanParsable<TSelf>`, which denotes that the sub-pattern  can be automatically parsed into a value of the designated type.
- A variable declaration cannot be `var` (there's no natural deconstruction type, as both char and string could be considered)

More examples of this alternative option:
```cs
metadataName is $"{string identifier}`{int arity}"

logLine is $"[{DateTime date}] {string logMessage}"
```

**Reply by kyoyama-kazusa** [reply]
IIRC this can be done by using interpolated string handlers like this:

<img width="2701" height="871" alt="image" src="https://github.com/user-attachments/assets/34ee37f3-b99f-4e03-9b11-c16532114337" />


A simple repro:

```c#
#pragma warning disable CS8500

using System;
using System.Collections.Generic;
using System.Text;
using System.Runtime.CompilerServices;
using System.Text.RegularExpressions;

string name = default!;
string age = default!;
var originalString = "Hello sensei. I'm Kazusa. I'm now 15 years old!";
VariableExtractor.Assign(
    // Original string text
    originalString,

    // Define an interpolated string in order to extract variables
    // (Just a simple implementation, don't allow any variables not of type string here)
    $"Hello sensei. I'm {name}. I'm now {age} years old!"
);
Console.WriteLine($"{nameof(name)} = \"{name}\", {nameof(age)} = {age}");


[InterpolatedStringHandler]
file ref struct VariableExtractor(int literalLength, int formattedCount, string original)
{
    /// <summary>
    /// Indicates the placeholder character. This character will be used as placeholders, just like
    /// <c>{0}</c> in <c>"Hello, {0}!"</c>.
    /// </summary>
    private const char PlaceholderReservedCharacter = '\0';


#pragma warning disable IDE0052
    /// <summary>
    /// Indicates the number of characters in literal.
    /// </summary>
    private readonly int _literalLength = literalLength;

    /// <summary>
    /// Indicates the number of interpolated items (variables).
    /// </summary>
    private readonly int _formattedCount = formattedCount;

    /// <summary>
    /// Indicates the original string to be parsed.
    /// </summary>
    private readonly string _original = original;
#pragma warning restore IDE0052

    /// <summary>
    /// Indicates the string variable pointers.
    /// </summary>
    private readonly unsafe string?*[] _pointers = new string*[formattedCount]!;

    /// <summary>
    /// Indicates the backing builder object.
    /// </summary>
    private readonly StringBuilder _builder = new();

    /// <summary>
    /// Indicates the current variable index.
    /// </summary>
    private int _index = 0;


    /// <summary>
    /// Appends a list of unchanged characters into the collection.
    /// </summary>
    /// <param name="s">The sequence of characters, unchanged.</param>
    public readonly void AppendLiteral(string s) => _builder.Append(s);

    /// <summary>
    /// Appends a variable into the collection, recording its variable pointer (address).
    /// </summary>
    /// <param name="variable">The reference to the variable.</param>
    public unsafe void AppendFormatted(in string? variable)
    {
        _builder.Append(PlaceholderReservedCharacter);
        _pointers[_index++] = (string?*)Unsafe.AsPointer(ref Unsafe.AsRef(in variable));
    }

    /// <inheritdoc cref="object.ToString"/>
    public readonly override string ToString()
        => _builder.ToString().Replace(PlaceholderReservedCharacter, '?');


    /// <summary>
    /// Try to analyze the text difference and assign to interpolated variables.
    /// </summary>
    /// <param name="original">Indicates the original text.</param>
    /// <param name="extractor">Indicates the interpolated string that inserts custom variables to be parsed.</param>
    /// <exception cref="InvalidOperationException">Throws when the text is invalid to be checked.</exception>
    public static unsafe void Assign(
        string original,
        [InterpolatedStringHandlerArgument(nameof(original))] in VariableExtractor extractor
    )
    {
        var newString = extractor._builder.ToString();
        var parts = newString.Split(PlaceholderReservedCharacter);
        if (parts.Length < 1)
        {
            return;
        }

        var regexParts = new List<string> { Regex.Escape(parts[0]) };
        for (var i = 1; i < parts.Length; i++)
        {
            regexParts.Add(/*lang = regex*/"(.*?)");
            regexParts.Add(Regex.Escape(parts[i]));
        }

        var match = Regex.Match(original, $"^{string.Concat(regexParts)}$");
        if (match is not { Success: true, Groups: var groups })
        {
            throw new InvalidOperationException("Failed to match string.");
        }

        for (var i = 1; i < groups.Count; i++)
        {
            // Assign string to property.
            Unsafe.AsRef<string?>(extractor._pointers[i - 1]) = groups[i].Value;
        }
    }
}
```

C# 10 introduces the feature "Interpolated String Handler Improvements", allowing us defining pattern-based interpolated string handlers, consuming methods `AppendText` and `AppendLiteral`. This feature also allows passing `in` parameters (by-reference read-only pararmeters) to do this.

Precisely because of this (by-ref parameter), we can do a "connection" between captured variables and internal references (`_pointers`).

😆 Don't abuse this. This is very strange 🤣

**Reply by jnm2** [reply]
Possibly also multiple string subpatterns could be allowed if their Length property is matched with a constant:
```
is $"{string { Length: 4 } prefix} - {string { Length: 6 } suffix}"
```
